### PR TITLE
Use github api for luda editor

### PR DESCRIPTION
**Pull request by bigfoodK**
# What I did
## Use github api instead of luda-editor-server
Read and write sequences and images with github api.
In this pr, user must enter the github authentication token to text input manually.
Authentication ui will be implemented in next pr.

I used my private repository as storage for develop.
Before merge this pr, It must be replaced by a repository of namseent.

## Remove luda-editor-server and luda-edior-nrpc
You need to run `git clean -f .` after merge this pr.

## TODO

- [ ] Decide which repository to use as a storage
- [ ] Decide which oauth client to use for oauth login

### CODE REVIEW

File: luda-editor/luda-editor-client/src/app/sequence_list/mod.rs [hunk 1]
```diff
@@ -165,3 +188,48 @@ fn calculate_sequence_duration(sequence: &Arc<Sequence>) -> Time {
...
+            }
+            Err(error) => namui::event::send(SequenceListEvent::SequenceOpenStateChangedEvent {
+                title: title.clone(),
+                state: SequenceOpenState::Failed {
+                    message: format!("{:#?}", error),
+                },
+            }),
+        }
+    });
+}
```
Comment (bigfoodK) [inline luda-editor/luda-editor-client/src/app/sequence_list/mod.rs:235]: Lock sequence first, then open editor

File: luda-editor/luda-editor-client/src/app/sequence_list/ops/sync_sequences_from_google_spreadsheet.rs [hunk 2]
```diff
@@ -53,37 +54,35 @@ impl SequenceList {
...
-        })
-        .await?;
+    for (title, sequence) in title_sequence_map {
+        storage
+            .put_sequence(title.as_str(), sequence)
+            .await
+            .map_err(|error| format!("failed to save sequence: {:#?}", error))?;
+    }
     Ok(())
 }
```
Comment (bigfoodK) [inline luda-editor/luda-editor-client/src/app/sequence_list/ops/sync_sequences_from_google_spreadsheet.rs:87]: save sequences one by one, not in parallel.
Conflict may occur when a write api is requested at the same time.
However, we don't have strategy to respond the conflict.
So we save the sequences  one by one.

File: luda-editor/luda-editor-client/src/app/storage/storage.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    + GithubStorageSequenceUnlock
+    + GithubStorageSequenceGet
+    + GithubStorageSequencePut
+    + GithubStorageSequenceListGet
+    + GithubStorageSequenceTitlesGet
+    + GithubStorageSequenceTitlesPut
+    + GithubStorageBackgroundImagePathsGet
+    + GithubStorageBackgroundImageGet
+    + GithubStorageCharacterImageGet
+    + GithubStorageCharacterImagePathsGet
```
Comment (bigfoodK) [inline luda-editor/luda-editor-client/src/app/storage/storage.rs:38]: Suuuuper long trait.
This is to divde the impls into several files while using one trait for easy mocking.

below is impossible for now.
``` rust
trait Trait {
  fn a();
  fn b();
}
struct Struct;

// a.rs
impl Trait for Struct {
  fn a() {}
}

// b.rs
impl Trait for Struct {
  fn b() {}
}
```

File: luda-editor/luda-editor-client/src/app/types/app_context.rs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+use super::{CameraAngleImageLoader, MetaContainer};
+use crate::app::storage::GithubStorage;
+use std::sync::Arc;
+
+pub struct AppContext {
+    pub storage: Arc<dyn GithubStorage>,
+    pub meta_container: Arc<MetaContainer>,
+    pub camera_angle_image_loader: Arc<dyn CameraAngleImageLoader>,
+}
```
Comment (bigfoodK) [inline luda-editor/luda-editor-client/src/app/types/app_context.rs:9]: `camera_angle_image_loader` is exactly same as `namui's image_loader`. But it uses `GithubStorage` instead.

File: luda-editor/luda-editor-client/src/app/authentication/authentication.rs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    const BASE_URL: &str = "https://api.github.com";
+    const OWNER: &str = "bigfoodK";
+    const REPO: &str = "api-test";
+    GithubAPiClient::new(
+        access_token,
+        BASE_URL.to_string(),
+        OWNER.to_string(),
+        REPO.to_string(),
+    )
+}
```
Comment (bigfoodK) [inline luda-editor/luda-editor-client/src/app/authentication/authentication.rs:104]: This is the part that needs to be replaced before it is merged.

Comment (bigfoodK) [reply luda-editor/luda-editor-client/src/app/authentication/authentication.rs:104]: Can I create a repository `luda-editor-storage` under namseent org?

**Comment by namse** [comment]
@bigfoodK If we decide to use namseent repository's branch for this system, what's the props and cons againt making other new repo?

**Comment by bigfoodK** [comment]
**pros**
We can use just one repo. I don't know if this will be adventage.

**cons**
This is just a resource, not a code and It will not be merged into master. So we won't take monorepo's merits.
While using this app a huge amount of commit history will be piled up. I think it's better to isolate this in another repo.

File: luda-editor/luda-editor-client/src/app/storage/storage.rs (hunk 3 above)
Comment (bigfoodK) [reply luda-editor/luda-editor-client/src/app/storage/storage.rs:38]: `StorageSequenceLockStateGet` name fixed.

Comment (bigfoodK) [reply luda-editor/luda-editor-client/src/app/storage/storage.rs:38]: Internal traits (GithubStorageMetaGet, ...) are not exposed outside GithubStorage.

**Comment by namse** [comment]
> This is just a resource, not a code and It will not be merged into master. So we won't take monorepo's merits.
I think it's a good merit because we can manage some resources without merging into master. If we make another repo, we have to manage new repository, so management point increases.

> While using this app a huge amount of commit history will be piled up. 
We can remove commits using github action interval jobs, kind of clean-up jobs.


Q. What about the permission of github? what kind of permission the user should have actually? Is it possible to divide the permission by branch?

**Comment by bigfoodK** [comment]
The user needs permission to push commit directly to the target branch.
GitHub does not seem to have a function to manage user permission by branch. However, there is a function that prevents committing directly to the branch.
What scenario do you have in mind?

**Comment by namse** [comment]
We can make a protected branch for this luda-editor and allow only some people to push their commit.
There is 'team' in github organization, so let's put editor people in that team.

**Comment by namse** [comment]
I think luda-editor developer needs to use another branch when they want to test codes. Is it possible?

**Review by namse — CHANGES_REQUESTED** [review]

**Comment by namse** [comment]
I think it's not a good idea to push this codes to master before making sure about git and repository strategy.
There are big problem about current concept.

If user A and B are trying to push new commit to master branch at the same time(or not very same time but if they fetched master branch with the same reference), they must make conflict even they try to write different files.
That's because branch only have 1 HEAD ref.

That mean, we maybe have to have each branch per each file. Then, it's not a good idea to use namseent repo for this because that will make too much of branches.

So, before making sure strategy, I suggest push this codes into different branch, not master. I think it's not ready. @bigfoodK

File: luda-editor/luda-editor-client/src/app/github_api/delete_file.rs [hunk 6]
```diff
@@ -0,0 +1,60 @@
+use super::{parse_response_as_json::ResponseParseError, types::RequestBuilder, GithubAPiClient};
+use serde::Serialize;
+
+impl GithubAPiClient {
```
[the remark this answers is not in the document]
Comment (bigfoodK) [reply luda-editor/luda-editor-client/src/app/github_api/delete_file.rs:4]: Thx!

File: luda-editor/luda-editor-client/src/app/authentication/authentication.rs [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    rect(RectParam {
+        rect: Rect::from_xy_wh(Xy::zero(), wh),
+        style: RectStyle {
+            stroke: None,
+            fill: Some(RectFill {
+                color: Color::grayscale_f01(0.3),
+            }),
+            round: None,
+        },
+    })
```
[the remark this answers is not in the document]
Comment (bigfoodK) [reply luda-editor/luda-editor-client/src/app/authentication/authentication.rs]: Oh, I'll try it.

**Comment by namse** [comment]
@bigfoodK  During this weekend, I considered many options and I think keep going with github as luda-editor's storage is a proper way to go.

So I made a new repository https://github.com/NamseEnt/luda-editor-storage for this.

I want to change the storage algorithm using other API like `git database api` for better usage.

Do you want to let me edit storage algorithm in PR now, or do you want to review in separated PR? I'll probably share the design documentation for the new algorithm first.

**Comment by namse** [comment]
PR Content Merged into #313 so I close this PR.

**Pull request closed without merging**